packages/jupyter-ai/jupyter_ai/context_providers/base.py

```diff
@@ -0,0 +1,211 @@
+            )
+            unified_parameters = {
+                "verbose": True,
+                **lm_provider_params,
+                **model_parameters,
+            }
+            llm = lm_provider(**unified_parameters)
+            self.llm = llm
+        return self.llm
+
+
+class ContextCommand(BaseModel):
+    cmd: str
+
+    @property
+    def id(self) -> str:
+        return self.cmd.partition(":")[0]
+
+    @property
+    def arg(self) -> Optional[str]:
+        if ":" not in self.cmd:
+            return None
+        return self.cmd.partition(":")[2].strip("'\"").replace("\\ ", " ")
+
+    def __str__(self) -> str:
+        return self.cmd
+
+    def __hash__(self) -> int:
+        return hash(self.cmd)
+
+
+class BaseCommandContextProvider(BaseContextProvider):
+    id_prefix: ClassVar[str] = "@"
+    only_start: ClassVar[bool] = False
+    requires_arg: ClassVar[bool] = False
+    remove_from_prompt: ClassVar[bool] = (
+        False  # whether the command should be removed from prompt
+    )
+
+    @property
+    def command_id(self) -> str:
+        return self.id_prefix + self.id
+
+    @property
+    def pattern(self) -> str:
+        # arg pattern allows for arguments between quotes or spaces with escape character ('\ ')
+        return (
+            rf"(?<![^\s.])@{self.id}:(?:'[^']+'|\"[^\"]+\"|[^\s\\]+(?:\\ [^\s\\]*)*)"
+            if self.requires_arg
+            else rf"(?<![^\s.])@{self.id}(?![^\s.])"
+        )
+
+    def replace_prompt(self, prompt: str) -> str:
+        """Cleans up commands from the prompt before sending it to the LLM"""
+
+        def replace(match):
+            if _is_within_backticks(match, prompt):
+                return match.group()
+            return self._replace_command(ContextCommand(cmd=match.group()))
+
+        return re.sub(self.pattern, replace, prompt)
+
+    def get_arg_options(self, arg_prefix: str) -> List[ListOptionsEntry]:
+        """Returns a list of autocomplete options for arguments to the command
+        based on the prefix.
+        Only triggered if ':' is present after the command id (e.g. '@file:').
+        """
+        if self.requires_arg:
+            # default implementation that should be modified if 'requires_arg' is True
+            return [self._make_arg_option(arg_prefix)]
+        return []
+
+    def _find_commands(self, text: str) -> List[ContextCommand]:
+        # finds commands of the context provider in the text
+        matches = list(re.finditer(self.pattern, text))
+        if self.only_start:
+            matches = [match for match in matches if match.start() == 0]
+        results = []
+        for match in matches:
+            if not _is_within_backticks(match, text):
+                results.append(ContextCommand(cmd=match.group()))
+        return results
+
+    def _replace_command(self, command: ContextCommand) -> str:
+        if self.remove_from_prompt:
+            return ""
+        return command.cmd
```

Comment:
(optional) It seems all installed context providers are called by default, and it is up to each context provider to determine if it should run. The problem is that a) this allows context providers to run on every query, b) these can be installed from third-party packages via the entry points API. This can be slow (from waiting on API calls before each query), expensive (from using more tokens per query), and dangerous (from running third-party code without the user's knowledge).

I think that in the initial implementation, every context provider should require an explicit command to be triggered. Perhaps in the future we can add a new `AutomaticContextProvider` class and require users to explicitly enable each one in Jupyter AI's settings somehow. However, in the initial implementation:

- The code for matching commands and removing them from the prompt should be defined outside of `BaseCommandContextProvider`, to prevent subclasses from overriding this logic.
- The implementation in `DefaultChatHandler` should use the code for matching commands to determine which context providers to call; it should not call all of them and trust them to determine when they should run.

This is a high-level design subject, so feel free to ask questions or discuss below in the comments. No action necessary if you have other ideas for how this should be designed.

Reply:
I did not consider this security perspective at all which is completely valid. 

I was only considering extendibility. The reason for matching being part of `BaseCommandContextProvider` is because general trigger detection is handled by `BaseContextProvider` to allow for other kinds of inferred triggering. E.g. such as via detecting presence of a specific field in the human message, as in the `ErrorContextProvider` example, or some other types of string inference based on the user prompt (e.g. some known internal subject keywords to trigger a specific retriever).

How about having some opt-in configuration in the UI? So installed third-party providers are disabled by default and have to be explicitly enabled by the user. I would need some help to design the UI/UX of such a config. We could always disable the ability to install third-party providers in the first release and only allow after such config is enabled. I'll find a way to hack in my custom providers for my use case in the meantime.

This is just what came to mind, I would need a while to think of some better ideas.

Reply:
For the initial implementation, can we restrict the scope to only consider context providers that are triggered via commands? I totally understand the use-case for automatically pulling in the error from the active cell, or for running the context provider intelligently based on a user's prompt through an LLM agent. However, these "automatic" context providers are sufficiently different from "command-driven" context providers that they can be implemented separately in a future PR.

In summary, I'd like to keep the scope of this PR narrow so it has a clear path to being merged in the near future. Let me know what you think about this.

Reply:
How about this
1. I filter the installed objects to only add `BaseCommandContextProvider` to the "jai_contenxt_providers".
2. I mark `_BaseContextProvider` as private.
3. I'll change the `make_context_prompt` abstractmethod, that the user has to implement, to take in the list of command instances and produce a context prompt. The logic for ignoring messages without command/s will be implemented in the base class itself and does not need to be handled by the user. This is a good change indepedent of this concern.